### PR TITLE
Fix/import imp

### PR DESCRIPTION
**Pull request by dendisuhubdy**
changing imp to importlib for Python>=3.4 compatibility

### CODE REVIEW

**Comment by naupaka** [comment]
+1 to fix this

**Comment by diraol** [comment]
@dendisuhubdy will this work on python 2.7 ?
If not, can you please do a try/except in order to keep it compatible?

**Comment by dendisuhubdy** [comment]
@diraol yep I need to add that

**Comment by diraol** [comment]
@dendisuhubdy can you take a look at this comment to see if it works as a fix for us too?

https://github.com/SirVer/ultisnips/issues/996#issuecomment-403347577

**Comment by dendisuhubdy** [comment]
@diraol I'd check it out tonight, can you in the meanwhile figure out the try exceptions? I think I can only push those during the weekends

**Comment by diraol** [comment]
Hi @dendisuhubdy What I would do is:
instead of changing `from imp import find_module` with `from importlib import find_module`, I would do:

```python
try:
    from imp import find_module
except ImportError:
    from importlib import find_module
```

Or the opposite.

**Comment by bkbncn** [comment]
Vim has already fixed this issue at patch 8.1.0201 vim/vim@79a494d5e2f97c10e74f92ea529552623c314422

After that, vim won't call imp itself, but some plugins still call imp.

That 'silent' workaround doesn't fix anything, just forced to skip warning message.

**Comment by freeyoung** [comment]
This pull request can NOT work.

```
Python 3.7.0 (default, Jun 29 2018, 20:13:13)
[Clang 9.1.0 (clang-902.0.39.2)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> from importlib import find_module
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ImportError: cannot import name 'find_module' from 'importlib' (/usr/local/Cellar/python/3.7.0/Frameworks/Python.framework/Versions/3.7/lib/python3.7/importlib/__init__.py)
>>>
```

**Comment by dendisuhubdy** [comment]
@freeyoung I'm figuring it out

**Comment by ryan-lane** [comment]
Any update or workaround for this? This is making life difficult in the homebrew world :(

**Comment by diraol** [comment]
Just a fast contribution to the discussion:
FROM: https://docs.python.org/3.7/library/imp.html#imp.find_module

> Deprecated since version 3.3: Use importlib.util.find_spec() instead unless Python 3.3 compatibility is required, in which case use importlib.find_loader(). For example usage of the former case, see the Examples section of the importlib documentation.

**Comment by diraol** [comment]
I still think we need a PY2/PY3 check (maybe using `six` for that) in order to support python 2.7

File: autoload/pymode/breakpoint.vim [hunk 1]
```diff
@@ -9,7 +9,7 @@ fun! pymode#breakpoint#init() "{{{
 
         PymodePython << EOF
 
-from imp import find_module
+from importlib import find_module
```
Comment (thomwiggers) [inline autoload/pymode/breakpoint.vim]: This is not possible, as many people pointed out, because `importlib.find_module` does not exist.

For this file, you can use  `importlib.utils.find_spec`. You can use that to write the following:
```python
from importlib.util import find_spec

for module in ('wdb', 'pudb', 'ipdb', 'pdb'):
    mod_ = find_spec(module)
    if mod_ is not None:
        vim.command('let g:pymode_breakpoint_cmd = "import %s; %s.set_trace()  # XXX BREAKPOINT"' % (module, module))
        break
```

To preserve python 2 / old python 3 compatibility, you may want to write a wrapper around `imp.find_module` so it behaves a bit like `find_spec`.

```python
def find_spec(name):
    try:
        from importlib.util import find_spec
        if find_spec(name) is not None:
            return name
    except ImportError:
        try:
            from imp import find_module
            find_module(name)
            return name
        except ImportError:
            return None
````

**Review by thomwiggers — CHANGES_REQUESTED** [review]

File: pymode/libs/logilab-common-1.4.1/logilab/common/modutils.py [hunk 2]
```diff
@@ -34,7 +34,7 @@
 import os
 from os.path import (splitext, join, abspath, isdir, dirname, exists,
                      basename, expanduser, normcase, realpath)
-from imp import find_module, load_module, C_BUILTIN, PY_COMPILED, PKG_DIRECTORY
+from importlib import find_module, load_module, C_BUILTIN, PY_COMPILED, PKG_DIRECTORY
```
Comment (thomwiggers) [inline pymode/libs/logilab-common-1.4.1/logilab/common/modutils.py]: `importlib` does not have `load_module` or `find_module`. Perhaps you could use `importlib.import_module`. It would probably make more sense to get this fixed upstream.

File: pymode/libs/six.py [hunk 3]
```diff
@@ -235,7 +235,7 @@ class _MovedItems(_LazyModule):
     MovedAttribute("intern", "__builtin__", "sys"),
     MovedAttribute("map", "itertools", "builtins", "imap", "map"),
     MovedAttribute("range", "__builtin__", "builtins", "xrange", "range"),
-    MovedAttribute("reload_module", "__builtin__", "imp", "reload"),
+    MovedAttribute("reload_module", "__builtin__", "importlib", "reload"),
```
Comment (thomwiggers) [inline pymode/libs/six.py]: Updating `six.py` from upstream would probably make more sense. It only uses importlib in python 3.4+: https://bitbucket.org/gutworth/six/src/92e1c746e1abfbdb94705b821bc93766083efc54/six.py?at=default&fileviewer=file-view-default#six.py-245

File: pymode/libs/pkg_resources/__init__.py [hunk 4]
```diff
@@ -43,7 +43,7 @@
     import _imp
 except ImportError:
     # Python 3.2 compatibility
-    import imp as _imp
+    import importlib as _imp
```
Comment (thomwiggers) [inline pymode/libs/pkg_resources/__init__.py]: `importlib` does not provide `_imp`. In fact, `importlib` uses `_imp`.

```
help(_imp)
Help on built-in module _imp:

NAME
    _imp - (Extremely) low-level import machinery bits as used by importlib and imp.

```

**Review by thomwiggers — CHANGES_REQUESTED** [review]

**Comment by ricardodani** [comment]
+1 to this

File: pymode/libs/pkg_resources/__init__.py (hunk 4 above)
Comment (dendisuhubdy) [reply pymode/libs/pkg_resources/__init__.py]: would fix it, just had time! Sorry about that

File: pymode/libs/six.py (hunk 3 above)
Comment (dendisuhubdy) [reply pymode/libs/six.py]: would fix it, just had time! Sorry about that

File: pymode/libs/logilab-common-1.4.1/logilab/common/modutils.py (hunk 2 above)
Comment (dendisuhubdy) [reply pymode/libs/logilab-common-1.4.1/logilab/common/modutils.py]: would fix it, just had time! Sorry about that

File: autoload/pymode/breakpoint.vim (hunk 1 above)
Comment (dendisuhubdy) [reply autoload/pymode/breakpoint.vim]: would fix it, just had time! Sorry about that

File: pymode/libs/logilab-common-1.4.1/logilab/common/compat.py [hunk 5]
```diff
@@ -68,7 +68,10 @@ def method_type(callable, instance, klass):
     reload = reload
 else:
     from io import FileIO, BytesIO, StringIO
-    from imp import reload
+    try:
+        from imp import reload
+    except ImportError:
+        from importlib import reload
```
Comment (thomwiggers) [inline pymode/libs/logilab-common-1.4.1/logilab/common/compat.py:74]: You'll want to do this the other way around; `from imp import reload` will work for now but raise a deprecation error.

File: pymode/libs/logilab-common-1.4.1/logilab/common/modutils.py [hunk 6]
```diff
@@ -34,10 +34,13 @@
...
 from os.path import (splitext, join, abspath, isdir, dirname, exists,
                      basename, expanduser, normcase, realpath)
-from imp import find_module, load_module, C_BUILTIN, PY_COMPILED, PKG_DIRECTORY
 from distutils.sysconfig import get_config_var, get_python_lib, get_python_version
 from distutils.errors import DistutilsPlatformError
 
+try:
+    from imp import find_module, load_module, C_BUILTIN, PY_COMPILED, PKG_DIRECTORY
+except:
+    from importlib import import_module, C_BUILTIN, PY_COMPILED, PKG_DIRECTORY
```
Comment (thomwiggers) [inline pymode/libs/logilab-common-1.4.1/logilab/common/modutils.py:43]:
```
>>> from importlib import import_module, C_BUILTIN, PY_COMPILED, PKG_DIRECTORY
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ImportError: cannot import name 'C_BUILTIN' from 'importlib' (/usr/lib/python3.7/importlib/__init__.py)
```

File: pymode/libs/logilab-common-1.4.1/logilab/common/modutils.py [hunk 7]
```diff
@@ -671,24 +676,27 @@ def _module_file(modpath, path=None):
...
             return C_BUILTIN, None
     imported = []
     while modpath:
         modname = modpath[0]
-        # take care to changes in find_module implementation wrt builtin modules
+        # take care to changes in find_module/find_spec implementation wrt builtin modules
         #
         # Python 2.6.6 (r266:84292, Sep 11 2012, 08:34:23)
-        # >>> imp.find_module('posix')
+        # >>> importlib.find_module('posix')
```
Comment (thomwiggers) [inline pymode/libs/logilab-common-1.4.1/logilab/common/modutils.py:689]: `find_module` does not exist in importlib.

File: pymode/libs/logilab-common-1.4.1/logilab/common/modutils.py [hunk 8]
```diff
@@ -579,7 +584,7 @@ def is_relative(modname, from_file):
     if from_file in sys.path:
         return False
     try:
-        find_module(modname.split('.')[0], [from_file])
+        find_spec(modname.split('.')[0], [from_file])
```
Comment (thomwiggers) [inline pymode/libs/logilab-common-1.4.1/logilab/common/modutils.py:587]: `find_spec` returns None if it can't find it, unlike `find_module` which crashes.

File: pymode/libs/logilab-common-1.4.1/logilab/common/modutils.py [hunk 9]
```diff
@@ -671,24 +676,27 @@ def _module_file(modpath, path=None):
...
         # Python 3.3.1 (default, Apr 26 2013, 12:08:46)
-        # >>> imp.find_module('posix')
+        # >>> importlib.find_module('posix')
+        # (None, None, ('', '', 6))
+        # Python 3.40 (default, Apr 26 2013, 12:08:46)
+        # >>> importlib.find_spec('posix')
         # (None, None, ('', '', 6))
         try:
-            _, mp_filename, mp_desc = find_module(modname, path)
+            _, mp_filename, mp_desc = find_spec(modname, path)
```
Comment (thomwiggers) [inline pymode/libs/logilab-common-1.4.1/logilab/common/modutils.py:699]: ```TypeError: cannot unpack non-iterable ModuleSpec object```

File: pymode/libs/six.py [hunk 10]
```diff
@@ -1 +0,0 @@
-../../submodules/six/six.py
```
Comment (thomwiggers) [inline pymode/libs/six.py:1]: Why remove this symlink instead of replacing `pymode/libs/six.py`?

**Review by thomwiggers — CHANGES_REQUESTED** [review]
There are still some big issues with this PR.

File: pymode/libs/logilab-common-1.4.1/logilab/common/modutils.py (hunk 9 above)
Comment (diraol) [reply pymode/libs/logilab-common-1.4.1/logilab/common/modutils.py:699]: `find_module` and `find_spec` are not a replacement for each other, they work in completely different ways.

**Comment by blayz3r** [comment]
+1 to fix this

**Comment by shiyuge** [comment]
would love to see this fixed

**Comment by diraol** [comment]
Folks, we need help here, this PR is not yet ready to be merged, so any contributions are VERY welcome! I haven't had enough time to work on this problem...

**Comment by dendisuhubdy** [comment]
@diraol hi man, I'll get contacted with you via email to discuss it?

**Comment by diraol** [comment]
I think that #950 fixed this, so I'm closing this PR. =)

**Comment by pazz** [comment]
I am running into #937 which was closed by referencing this issue. This issue here was closed because #950 got merged. However, I *still* see the initially reported error, both on the current "develop" and "issue_971_import_machinery" branch. Please reopen #937 or advice on how to fix the issue.

Python 3.7.2
VIM - Vi IMproved 8.1 (2018 May 18, compiled Jan 05 2019 20:53:12) with +python3
both from Debian testing

**Comment by lalmeras** [comment]
[not a python-mode developer, just user]

@pazz, you may follow #972 that is the issue following this breakage. It seems this issue is really tricky to fix, as any change working in a given context breaks usage for other contexts.

I post a message with a fix that works for me (based on branch issue_971_import_machinery) on this issue. From my side, I suspect that the issue is related to vim builds with both python and python3 enabled (+python and +python3), but I do not see any feedback confirming or invalidating it.

TLDR; this issue is still not fixed and tricky, please see #972 for updates.

**Comment by pazz** [comment]
Thanks for your quick reply.
For what it's worth, my vim build has +python3 and -python.
